lib/groonga-query-log/command/check-performance-regression.rb

```diff
@@ -0,0 +1,329 @@
+          old_elapsed_nsec = statistic[:old_elapsed_nsec]
+          new_elapsed_nsec = statistic[:new_elapsed_nsec]
+
+          if slow_response?(old_elapsed_nsec, new_elapsed_nsec)
+            @output.puts("Query: #{query}")
+            ratio = statistic[:ratio]
+            @output.puts("  %s" % [
+              format_elapsed_calculated_ratio(ratio, old_elapsed_nsec, new_elapsed_nsec)
+            ])
+            @output.puts("  Operations:")
+            old_operation_nsecs = average_elapsed_operation_nsecs(old_queries[query])
+            new_operation_nsecs = average_elapsed_operation_nsecs(new_queries[query])
+            old_operation_nsecs.each_with_index do |operation, index|
+              new_operation = new_operation_nsecs[index]
+              if slow_operation?(operation[:elapsed], new_operation[:elapsed])
+                @output.puts("    Operation: %s %s" % [
+                  operation[:name],
+                  format_elapsed_ratio(operation[:elapsed],
+                                       new_operation[:elapsed])
+                ])
+              end
+            end
+          end
+        end
+
+        if @output.kind_of?(File)
+          @output.close
+        end
+
+        true
+      end
+
+      private
+      def elapsed_ratio(old_elapsed_nsec, new_elapsed_nsec)
+        if old_elapsed_nsec == 0 and new_elapsed_nsec == 0
+          0.0
+        elsif old_elapsed_nsec == 0 and new_elapsed_nsec > 0
+          if new_elapsed_nsec / NSEC_IN_SECONDS < @options[:slow_response_threshold]
+            -Float::INFINITY
+          else
+            Float::INFINITY
+          end
+        else
+          (new_elapsed_nsec / old_elapsed_nsec) * 100 - 100
+        end
+      end
+
+      def average_elapsed_nsec(statistics)
+        elapsed_times = statistics.collect do |statistic|
+          statistic.elapsed
+        end
+        elapsed_times.inject(:+).to_f / elapsed_times.size
+      end
+
+      def average_elapsed_operation_nsecs(statistics)
+        operations = []
+        statistics.first.operations.each_with_index do |operation, index|
+          elapsed_times = statistics.collect do |statistic|
+            statistic.operations[index][:relative_elapsed]
+          end
+          operations << {
+            :name => statistics.first.operations[index][:name],
+            :elapsed => elapsed_times.inject(:+).to_f / elapsed_times.size
+          }
+        end
+        operations
+      end
+
+      def slow_response?(old_elapsed_nsec, new_elapsed_nsec)
+        ratio = elapsed_ratio(old_elapsed_nsec, new_elapsed_nsec)
+        elapsed_sec = ((new_elapsed_nsec - old_elapsed_nsec) / NSEC_IN_SECONDS)
+        slow_response = ((ratio >= @options[:slow_response_ratio]) and
+                        (elapsed_sec >= @options[:slow_response_threshold]))
+        slow_response
+      end
+
+      def slow_operation?(old_elapsed_nsec, new_elapsed_nsec)
+        ratio = new_elapsed_nsec / old_elapsed_nsec * 100 - 100
```

Comment:
Don't we have to avoid division by zero here?

Reply:
Fixed to use already supported elapsed_ratio